http-clients/src/main/java/com/palantir/remoting/http/GuavaOptionalAwareContract.java

```diff
@@ -39,50 +35,32 @@
  * <p>
  * {@link PathParam}s require a value, and so we explicitly disallow use with {@link Optional}.
  */
-public final class GuavaOptionalAwareContract implements Contract {
-
-    private final Contract delegate;
+public final class GuavaOptionalAwareContract extends AbstractDelegatingContract {
 
     public GuavaOptionalAwareContract(Contract delegate) {
-        this.delegate = delegate;
+        super(delegate);
     }
 
     @Override
-    public List<MethodMetadata> parseAndValidatateMetadata(Class<?> targetType) {
-        List<MethodMetadata> mdList = delegate.parseAndValidatateMetadata(targetType);
-        Map<String, MethodMetadata> methodMetadataByConfigKey = new LinkedHashMap<String, MethodMetadata>();
-        for (MethodMetadata md : mdList) {
-            methodMetadataByConfigKey.put(md.configKey(), md);
-        }
-
-        for (Method method : targetType.getMethods()) {
-            if (method.getDeclaringClass() == Object.class) {
-                continue;
-            }
-            String configKey = Feign.configKey(targetType, method);
-            MethodMetadata md = methodMetadataByConfigKey.get(configKey);
-            if (md != null) {
-                Class<?>[] parameterTypes = method.getParameterTypes();
-                Annotation[][] annotations = method.getParameterAnnotations();
-                for (int i = 0; i < parameterTypes.length; i++) {
-                    Class<?> cls = parameterTypes[i];
-                    if (cls.equals(Optional.class)) {
-                        FluentIterable<Class<?>> paramAnnotations =
-                                FluentIterable.from(Lists.newArrayList(annotations[i])).transform(EXTRACT_CLASS);
-                        if (paramAnnotations.contains(HeaderParam.class)) {
-                            md.indexToExpanderClass().put(i, GuavaEmptyOptionalExpander.class);
-                        } else if (paramAnnotations.contains(QueryParam.class)) {
-                            md.indexToExpanderClass().put(i, GuavaNullOptionalExpander.class);
-                        } else if (paramAnnotations.contains(PathParam.class)) {
-                            throw new RuntimeException(String.format(
-                                    "Cannot use Guava Optionals with PathParams. (Class: %s, Method: %s, Param: arg%d)",
-                                    targetType.getName(), method.getName(), i));
-                        }
-                    }
+    protected void processMetadata(Class<?> targetType, Method method, MethodMetadata metadata) {
```

Comment:
Yes, pushes up common structure/functionality to `AbstractDelegatingContract` and then just re-writes to take advantage of that and reduce duplicated code -- no functionality change.